rpc/src/history.rs

```diff
@@ -0,0 +1,206 @@
+use crate::jsonrpsee::core::{async_trait, Error, RpcResult};
+use anyhow::anyhow;
+use ethportal_api::types::discv5::{Enr, NodeId, RoutingTableInfo};
+use ethportal_api::types::portal::{
+    AcceptInfo, ContentInfo, DataRadius, FindNodesInfo, PaginateLocalContentInfo, PongInfo,
+    TraceContentInfo,
+};
+use ethportal_api::HistoryNetworkApiServer;
+use ethportal_api::{HistoryContentItem, HistoryContentKey};
+
+use ethportal_api::jsonrpsee::core::__reexports::serde_json::from_value;
+use ethportal_api::jsonrpsee::core::__reexports::serde_json::Value;
+use tokio::sync::mpsc;
+use trin_core::jsonrpc::endpoints::HistoryEndpoint;
+use trin_core::jsonrpc::types::HistoryJsonRpcRequest;
+
+pub struct HistoryNetworkApi {
+    network: mpsc::UnboundedSender<HistoryJsonRpcRequest>,
+}
+
+impl HistoryNetworkApi {
+    pub fn new(network: mpsc::UnboundedSender<HistoryJsonRpcRequest>) -> Self {
+        Self { network }
+    }
+
+    pub async fn proxy_query_to_history_subnet(
+        &self,
+        endpoint: HistoryEndpoint,
+    ) -> anyhow::Result<Value> {
+        let (resp_tx, mut resp_rx) = mpsc::unbounded_channel::<Result<Value, String>>();
+        let message = HistoryJsonRpcRequest {
+            endpoint,
+            resp: resp_tx,
+        };
+        let _ = self.network.send(message);
+
+        match resp_rx.recv().await {
+            Some(val) => match val {
+                Ok(result) => Ok(result),
+                Err(msg) => Err(anyhow!(msg)),
+            },
+            None => Err(anyhow!(
+                "Internal error: No response from chain history subnetwork"
+            )),
+        }
+    }
+}
+
+#[async_trait]
+impl HistoryNetworkApiServer for HistoryNetworkApi {
+    /// Returns meta information about overlay routing table.
+    async fn routing_table_info(&self) -> RpcResult<RoutingTableInfo> {
+        let endpoint = HistoryEndpoint::RoutingTableInfo;
+        let result = self.proxy_query_to_history_subnet(endpoint).await?;
+        let result: RoutingTableInfo = from_value(result)?;
+        Ok(result)
+    }
+
+    /// Write an Ethereum Node Record to the overlay routing table.
+    async fn add_enr(&self, _enr: Enr) -> RpcResult<bool> {
+        Err(Error::MethodNotFound("add_enr".to_owned()))
```

Comment:
When I started with `ethportal-api`, I built the API traits following the portal json-rpc specs. One reason for this is the Client trait which should be compatible with any portal client, not only Trin.

In Trin, we don't currently implement all rpc endpoints from the specs, so I think it is acceptable to return a MethodNotFound error but keep the `ethportal-api` traits because Ultralight for example implement some of what is missing in Trin and a jsonrpsee client may want to interact with such a node.